### PR TITLE
Cuda sticky allocator

### PR DESCRIPTION
**Pull request by abadams**
Opt-in allocation cache for cuda. Speeds up local laplacian and bilateral grid by 3x!

I don't have a good eviction policy yet though. The current limit is just a max of 32 unused allocations, regardless of their size. I tried a high-water mark of memory usage so far, but it was too aggressive at freeing memory, and isn't great for a long-running process anyway. Suggestions welcome.

### CODE REVIEW

File: src/runtime/cuda.cpp [hunk 1]
```diff
@@ -136,7 +147,7 @@ WEAK int halide_cuda_acquire_context(void *user_context, CUcontext *ctx, bool cr
         }
 
         {
-            ScopedSpinLock spinlock(&context_lock);
+            ScopedMutexLock spinlock(&context_lock);
```
Comment (dsharletg) [inline src/runtime/cuda.cpp]: @zvookin might be interested in this change

File: src/runtime/cuda.cpp [hunk 2]
```diff
@@ -211,11 +222,10 @@ class Context {
...
     INLINE ~Context() {
-        if (error == 0) {
-            CUcontext old;
-            cuCtxPopCurrent(&old);
-            halide_cuda_release_context(user_context);
-        }
+        CUcontext old;
+        cuCtxPopCurrent(&old);
+
+        halide_cuda_release_context(user_context);
```
Comment (dsharletg) [inline src/runtime/cuda.cpp]: I think only this should come out of the `if (error == 0)`

File: src/runtime/cuda.cpp [hunk 3]
```diff
@@ -552,19 +605,48 @@ WEAK int halide_cuda_device_free(void *user_context, halide_buffer_t* buf) {
...
     uint64_t t_before = halide_current_time_ns(user_context);
     #endif
 
     halide_assert(user_context, validate_device_pointer(user_context, buf));
 
-    debug(user_context) <<  "    cuMemFree " << (void *)(dev_ptr) << "\n";
-    CUresult err = cuMemFree(dev_ptr);
-    // If cuMemFree fails, it isn't likely to succeed later, so just drop
-    // the reference.
+    CUresult err = CUDA_SUCCESS;
```
Comment (dsharletg) [inline src/runtime/cuda.cpp]: If possible, I think this should be in a helper function wrapper around cuMemAlloc/cuMemFree, even if this helper function is not something that is overrideable a custom runtime (re: our gitter conversation about this).

File: src/runtime/cuda.cpp [hunk 4]
```diff
@@ -688,16 +773,85 @@ WEAK int halide_cuda_device_malloc(void *user_context, halide_buffer_t *buf) {
...
-    debug(user_context) << "    cuMemAlloc " << (uint64_t)size << " -> ";
-    CUresult err = cuMemAlloc(&p, size);
-    if (err != CUDA_SUCCESS) {
-        debug(user_context) << get_error_name(err) << "\n";
-        error(user_context) << "CUDA: cuMemAlloc failed: "
-                            << get_error_name(err);
-        return err;
-    } else {
-        debug(user_context) << (void *)p << "\n";
+    CUdeviceptr p = 0;
```
Comment (dsharletg) [inline src/runtime/cuda.cpp]: Same comment as on the free side.

File: src/runtime/cuda.cpp [hunk 5]
```diff
@@ -669,7 +754,7 @@ WEAK int halide_cuda_device_malloc(void *user_context, halide_buffer_t *buf) {
         return ctx.error;
     }
 
-    size_t size = buf->size_in_bytes();
+    size_t size = quantize_allocation_size(buf->size_in_bytes());
```
Comment (dsharletg) [inline src/runtime/cuda.cpp]: Only want to do this if caching allocations

File: src/runtime/cuda_functions.h [hunk 6]
```diff
@@ -56,6 +56,7 @@ CUDA_FN_4000(CUresult, cuCtxPopCurrent, cuCtxPopCurrent_v2, (CUcontext *pctx));
 CUDA_FN(CUresult, cuPointerGetAttribute, (void *result, int query, CUdeviceptr ptr));
 
 CUDA_FN_OPTIONAL(CUresult, cuStreamSynchronize, (CUstream hStream));
+CUDA_FN_OPTIONAL(CUresult, cuLaunchHostFunc, (CUstream hStream, void (*fn)(void *), void *));
```
Comment (dsharletg) [inline src/runtime/cuda_functions.h]: Don't see this referenced anywhere else in this PR?

**Comment by ongjunjie** [comment]
I'm watching this PR with great interest. I'm also seeing 3x improvements on some of my filters :)

Some feedback on JIT usage:
- Having to call an internal function is a little weird
- It's not immediately obvious that you must have JIT compiled the shared runtime before calling `halide_reuse_device_allocations()`

**Comment by steven-johnson** [comment]
Given the 3x performance increase in some cases, clearly this PR is still interesting; is the lack of an good eviction policy the gating factor?

**Comment by abadams** [comment]
Yes, on one hand the performance win seems like a must-have, but on the other this just leaves memory allocated indefinitely, and I don't have a good fix for that. There has to be something more clever we could do. E.g. somehow enqueing a callback so that we start freeing things when Halide hasn't used the GPU for a while.

**Comment by abadams** [comment]
In the absence of something cleverer, I think we should merge this as-is. I think I hit the outstanding comments. I just encountered a case where someone concluded Halide was bad for cuda. Their benchmark was spending a lot of its time doing malloc/free.

**Comment by steven-johnson** [comment]
SGTM, but can we go ahead and open an issue to track "may need good eviction policy", and annotate in the relevant chunk of code?

**Comment by abadams** [comment]
#4093

**Pull request merged**